torch/cuda/__init__.py

```diff
@@ -209,22 +209,31 @@ class device(object):
             this argument is a negative integer or ``None``.
     """
 
-    def __init__(self, device):
+    def __init__(self, device: Any):
         self.idx = _get_device_index(device, optional=True)
         self.prev_idx = -1
 
     def __enter__(self):
         if self.idx == -1:
             return
-        self.prev_idx = torch._C._cuda_getDevice()
+        self.prev_idx = torch.cuda.current_device()
         if self.prev_idx != self.idx:
-            torch._C._cuda_setDevice(self.idx)
-        _lazy_init()
-
-    def __exit__(self, *args):
+            # Eager and script mode API to set the device are different.
+            # Call the corresponding setDevice API depending on the mode
+            if torch.jit.is_scripting():
+                torch.cuda._setDevice(self.idx)
+            else:
+                torch._C._cuda_setDevice(self.idx)
+                _lazy_init()
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
         if self.prev_idx != self.idx:
-            torch._C._cuda_setDevice(self.prev_idx)
-        return False
+            # Eager and script mode API to re-set the device are different.
+            # Call the corresponding setDevice API depending on the mode.
```

Comment:
If there is a Note somewhere explaining why the eager and script mode APIs are different, that would be quite helpful

Reply:
I think it's because JIT doesn't "understand" `torch._C` since literally everything is in that extension module. I think it would be useful to introduce a layer of indirection here, say, `torch.cuda.setDevice` that calls `torch._C._cuda_setDevice` in eager mode and is desugared appropriately in JIT.

Reply:
(grep for `Note` to see existing examples of this)

Reply:
There is no note, but Eager and script API only differ at the API level but there is no change in the functionality.

Reply:
OK... well, if they only differ in API, why don't we just have one API (rather than having to switch between the two of them)? If JIT can't resolve torch._C, then an indirection as @SplitInfinity suggest seems better